src/components/pdf-viewer/layers/util/generateHighlights/index.ts

```diff
@@ -0,0 +1,106 @@
+import { SearchResults } from '../../../../../stores/search/PdfSearchStore'
+
+type Tuple = [number, number]
+
+interface HighlightsData {
+    matchedIndexes: number[]
+    tuples: Tuple[]
+}
+
+export const flattenDOMNodes = (nodes: HTMLCollection): any[] => {
+    const traverseNodes = (node: Node): any[] =>
+        node.nodeType === Node.ELEMENT_NODE ? [node as any, ...Array.from(node.childNodes).flatMap((child) => traverseNodes(child))] : []
+
+    return Array.from(nodes).flatMap((node) => traverseNodes(node))
+}
+
+export const generateHighlightsData = (nodes: any[], query: string): HighlightsData => {
+    const queryLowerCase = query?.toLowerCase()
+    const tuples: Tuple[] = []
+    const matchedIndexes: number[] = []
+    const textContent = nodes.reduce((prev, curr) => (prev += curr?.textContent?.toLowerCase() || ''), '')
+
+    let startIndex = textContent.indexOf(queryLowerCase)
+
+    while (startIndex !== -1) {
+        matchedIndexes.push(startIndex)
+        startIndex = String(textContent).indexOf(queryLowerCase, startIndex + 1)
+    }
+
+    nodes.forEach((node, i) => {
+        Array.from(node.textContent).forEach((_, j) => {
+            tuples.push([i, j])
+        })
+    })
+
+    return { matchedIndexes, tuples }
+}
+
+const getMarkedTextLength = (text: string): number => {
+    const markRegex = /<mark\b[^>]*>(.*?)<\/mark>/g
+    const markedText = text.match(markRegex)
+    if (markedText) {
+        return markedText.reduce((totalLength, mark) => totalLength + mark.length, 0)
+    }
+    return 0
+}
+
+export const generateHighlights = (
+    searchResults: SearchResults[],
+    container: Element,
+    query: string,
+    page: any,
+    currentHighlightIndex: number
+): any[] => {
+    if (!query) return []
+    let highlighIndex = 0
+    const currentPageSearchResults = searchResults.filter((match) => match.pageNum === page.pageNumber)
+    if (!currentPageSearchResults.length) return []
+
+    const nodes = flattenDOMNodes(container.children)
+    const { matchedIndexes, tuples } = generateHighlightsData(nodes, query)
+    if (!matchedIndexes.length) return []
+
+    const createMark = (text: string): string =>
+        `<mark ${currentPageSearchResults[highlighIndex]?.index === currentHighlightIndex ? 'class="active"' : ''} id="highlight-${
```

Comment:
see above; could we separate setting the orange highlight from creating and moving the mark around?